pineappl_capi/src/lib.rs

```diff
@@ -1357,3 +1355,253 @@ pub unsafe extern "C" fn pineappl_string_delete(string: *mut c_char) {
+/// `factors` with length of `combinations`.
+#[no_mangle]
+pub unsafe extern "C" fn pineappl_lumi_add2(
+    lumi: *mut Lumi,
+    combinations: usize,
+    nb_combinations: usize,
+    pdg_id_combinations: *const i32,
+    factors: *const f64,
+) {
+    let lumi = unsafe { &mut *lumi };
+    let pdg_id_pairs =
+        unsafe { slice::from_raw_parts(pdg_id_combinations, nb_combinations * combinations) };
+    let factors = if factors.is_null() {
+        vec![1.0; combinations]
+    } else {
+        unsafe { slice::from_raw_parts(factors, combinations) }.to_vec()
+    };
+
+    lumi.0.push(Channel::new(
+        pdg_id_pairs
+            .chunks(nb_combinations)
+            .zip(factors)
+            .map(|x| ((0..nb_combinations).map(|i| x.0[i]).collect(), x.1))
+            .collect(),
+    ));
+}
+
+/// Creates a new and empty grid that can accept any number of convolutions. The creation requires
+/// the following different sets of parameters:
+/// - The PID basis `pid_basis`: The basis onto which the partons are mapped, can be `Evol` or `Pdg`.
+/// - The channel function `channels`: A pointer to the luminosity function that specifies how the
+///   cross section should be reconstructed.
+/// - Order specification `orders` and `order_params`. Each `PineAPPL` grid contains a number of
+///   different perturbative orders, specified by `orders`. The array `order_params` stores the
+///   exponent of each perturbative order and must contain 4 integers denoting the exponent of the
+///   string coupling, of the electromagnetic coupling, of the logarithm of the renormalization
+///   scale, and finally of the logarithm of the factorization scale.
+/// - The observable definition `bins` and `bin_limits`. Each `PineAPPL` grid can store observables
+///   from a one-dimensional distribution. To this end `bins` specifies how many observables are
+///   stored and `bin_limits` must contain `bins + 1` entries denoting the left and right limit for
+///   each bin.
+/// - The types of convolutions `convolution_types` and their numbers `nb_convolutions`: specify how
+///   how many different convolutions are involved and their types - which are a cross product of the
+///   the following combination: (unpolarized, polarized) ⊗ (PDF, Fragmentation Function).
+/// - The PDG IDs of the involved initial- or final-state hadrons `pdg_ids`.
+/// - The types of kinematics `kinematics`: specify the various kinematics required to construct the
+///   Grid. These can be the energy and the various momentum fractions.
+/// - The specifications of the interpolation methods `interpolations`: provide the specifications on
+///   how each of the kinematics should be interpolated.
+///
+/// # Safety
+///
+/// # Panics
+#[no_mangle]
+#[must_use]
+pub unsafe extern "C" fn pineappl_grid_new2(
+    pid_basis: *const c_char, // takes: `const char* pid_basis = "Evol";`
+    channels: *const Lumi,
+    orders: usize,
+    order_params: *const u32,
```

Comment:
Let's use `u8` instead of `u32` for `order_params`, then you can get rid of the `.try_into().unwrap()` below.